### PR TITLE
Add errorSource to plugin metrics

### PR DESCRIPTION
**Pull request by LudoVio**
Add an `errorSource` label to plugins metrics. This is to help setting SLOs and Alert for plugin requests.

Similar to this:
https://docs.google.com/document/d/1_8KSW3O6R-0hzcVq9VPj9iLoB064NEdJDIEe0uh58Zo/edit

But for plugin request.

Logic:
  If plugin request return an error => errorSource = plugin
  If there is no error, then it depends on the different http status code returned by the database:
    - 500 => errorSource = database
    - 401, 402, 403, 407 => errorSource = external (could be config or infra)
    - 4xx => errorSource = plugin
   
If there is multiple responses from the database, with different errorSource, then we return the most sever one.
The priority is: pluginSource > databaseSource > externalSource > noneSource

That is the default implementation, it fits prometheus and probably others http databases. The next step would be to update plugin-sdk-go to add a way for plugins to override the default implementation.

### CODE REVIEW

**Comment by marefr** [comment]
Planning to have a look at this soon

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go [hunk 1]
```diff
@@ -58,29 +58,105 @@ const (
...
+	updateMetrics(pluginCtx.PluginID, endpointQueryData, string(cfg.Target), elapsed, status, string(errorSource))
+	pluginRequestSizeHistogram.WithLabelValues("grafana-backend", pluginCtx.PluginID, endpointQueryData, string(cfg.Target)).Observe(requestSize)
+	logDatasourceRequests(ctx, cfg, pluginCtx, endpointQueryData, status, elapsed, timeBeforePluginRequest, err)
+
+	return resp, err
+}
+
+func getErrorSource(status string, resp *backend.QueryDataResponse) errorSource {
+	if status == statusError {
+		return grafanaSource
```
Comment (marefr) [inline pkg/plugins/backendplugin/instrumentation/instrumentation.go:112]: Couldn't this be either Grafana or downstream?

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go [hunk 2]
```diff
@@ -58,29 +58,105 @@ const (
 	endpointQueryData      = "queryData"
 )
 
+type errorSource string
+
+const (
+	grafanaSource          errorSource = "grafana"
+	userSource             errorSource = "user"
+	downstreamClientSource errorSource = "downstreamClient"
+	downstreamServerSource errorSource = "downstreamServer"
```
Comment (marefr) [inline pkg/plugins/backendplugin/instrumentation/instrumentation.go:67]: I don't understand the use case of having these two divided.

**Review by marefr — CHANGES_REQUESTED** [review]
This looks okay, but it's a bit hard to follow and understand the logic. 

This doesn't specifically implement proposal in shared design doc because that refers to incoming HTTP request metrics, but it does introduce a similar `errorSource` label. Your use of the `errorSource` label does however include more values. We're planning to create an issue for the needed work for HTTP request metrics and `errorSource`. Maybe it would be easier to take this on after that one has been implemented?

**Comment by marefr** [comment]
Issue has been created https://github.com/grafana/grafana/issues/68480

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go [hunk 3]
```diff
@@ -58,29 +58,105 @@ const (
...
 	err := fn()
-	if err != nil {
-		status = statusError
-		if errors.Is(err, context.Canceled) {
-			status = statusCancelled
+
+	elapsed := time.Since(start)
+	status := getStatus(err)
+
+	logger.Info("Plugin Request Completed")
```
[the remark this answers is not in the document]
Comment (LudoVio) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:83]: > Now handled in `logDatasourceRequests`, isn't it?

Correct

No this PR is not an implementation of the proposal in the document. What I am trying to achieve here is "similar" in regard to the end goal, which is to have more information about where an error come from, without breaking the way errors are currently dealt with. But instead of incoming http request, here it's plugin request, so a go function call.
 
I don't see a dependency between this PR and the work on the HTTP request, but maybe I am wrong or you would prefere to have aligned implementations or error source for examples

[the remark this answers is not in the document]
Comment (marefr) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:83]: I think the dependency might be how you decide/propagate the source of an error. I think the general idea regarding /api/ds/query is
- If at least one of the query data responses contains an error the HTTP status returned is 400 Bad Request and errorSource=downstream
- If an error is returned from a plugin/QueryData the HTTP status returned is 500 Internal Server Error and errorSource=downstream if the error originating from plugin and not Grafana and the question/similarity with your PR vs our issue is how do we know that

Does this make more sense?

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go (hunk 1 above)
Comment (LudoVio) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:112]: it all depends of the plugins implementation I guess. For Prometheus, if this return an error, then it is something that happened before the request to the downstream. Then whatever the downstream response status code, no error are returned.
In that case it seems like if there is an error, then it is before downstream is involved.

Different plugins can have different implementation and behavior regarding when they return error or not.

The plugin-sdk interfaces seems to be limited here: error/ not error. 
Ideally we could pass some free form meta data too, so we can keep the current behavior for error, but let plugins returns meta data about where the error code from for example

File: pkg/tsdb/prometheus/querydata/request.go [hunk 4]
```diff
@@ -148,6 +153,9 @@ func (s *QueryData) fetch(ctx context.Context, client *client.Client, q *models.
 			logger.Error("Exemplar query failed", "query", q.Expr, "err", res.Error)
 		}
 		dr.Frames = append(dr.Frames, res.Frames...)
+		if res.Status > dr.Status {
+			dr.Status = res.Status
+		}
```
[the remark this answers is not in the document]
Comment (kylebrandt) [reply pkg/tsdb/prometheus/querydata/request.go:158]: We combine multiple queries to prometheus data source into one query object from Grafana. For example, including exemplars or doing "both" query option (which will hit the range and instant query endpoints).

For "both":
 - Remove it from dashboards
 - issue the range and instant queries in a separate query

For Exemplars:
 - I am less sure of the right thing to do here long term. But exemplars are kind of "extra" as I understand it, so I don't think we want to be careful about how we raise an warning/error here.

It seems the goal of the PR is primarily instrumentation, if that is that the I think we shouldn't be changing error return behavior and should address that separately.

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go (hunk 2 above)
Comment (LudoVio) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:67]: "downstream" is when downstream return anything else than http response code 200.
I created this two to make the differences between the "4xx" errors, which are most likely due to the user (client), like auth, and the "5xx" errors, which should never happen and are real downstream error

I don't have any strong opinion on this, what do you suggest?

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (LudoVio) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:83]: Yes thanks

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go (hunk 1 above)
Comment (bergquist) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:112]: I think it would be reasonable to add something to the plugin-sdk about where the error came from. It's fine if it does not work for all current plugins. The goal here is to improve the monitoring for the datasources we care the most about and those would be updated quickly. 

@marefr WDYT?

Comment (marefr) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go]: @bergquist I'm not sure that make sense. We already allow plugins to define their own metrics, wonder if that being the way forward given different plugins might have different needs. Do you have any suggestion of what that might look like?

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (marefr) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:82]: The log statement `logger.Info("Plugin Request Completed")` is already in `logDatasourceRequest` which you call further down in this function so is it really needed here as well? If so, why?
https://github.com/grafana/grafana/blob/6d8eac38bdf7286398b62cf7b25b68e49d1062d1/pkg/plugins/backendplugin/instrumentation/instrumentation.go#L85
https://github.com/grafana/grafana/blob/6d8eac38bdf7286398b62cf7b25b68e49d1062d1/pkg/plugins/backendplugin/instrumentation/instrumentation.go#L210

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go [hunk 5]
```diff
@@ -58,29 +58,127 @@ const (
...
+		return databaseSource
+	}
+
+	if res.Status >= 400 {
+		// Those error codes are related to authentication and authorization.
+		if res.Status == 401 || res.Status == 402 || res.Status == 403 || res.Status == 407 {
+			return externalSource
+		}
+
+		return pluginSource
```
Comment (marefr) [inline pkg/plugins/backendplugin/instrumentation/instrumentation.go:124]: Isn't a 400 a client/external source? Client has sent a bad request that plugin couldn't handle.

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go [hunk 6]
```diff
@@ -58,29 +58,127 @@ const (
...
+	errorSource := getErrorSource(status, resp)
+
+	updateMetrics(pluginCtx.PluginID, endpointQueryData, string(cfg.Target), elapsed, status, string(errorSource))
+	pluginRequestSizeHistogram.WithLabelValues("grafana-backend", pluginCtx.PluginID, endpointQueryData, string(cfg.Target)).Observe(requestSize)
+	logDatasourceRequests(ctx, cfg, pluginCtx, endpointQueryData, status, elapsed, timeBeforePluginRequest, err)
+
+	return resp, err
+}
+
+func getErrorSourceForResponse(res backend.DataResponse) errorSource {
```
Comment (marefr) [inline pkg/plugins/backendplugin/instrumentation/instrumentation.go:109]: Looking at this I keep coming back to whether you want to track this per request with the existing metrics when we treat request/response as a single "entity". Maybe what you're looking for would be a data_query_response metric of some kind that can track each one rather than trying to fit this into the existing metrics?

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go [hunk 7]
```diff
@@ -58,29 +58,127 @@ const (
...
+	if res.Error != nil {
+		return pluginSource
+	}
+
+	if res.Status >= 500 {
+		return databaseSource
```
Comment (marefr) [inline pkg/plugins/backendplugin/instrumentation/instrumentation.go:115]: Database in this case means the external system the plugin integrates with? This doesn't look bulletproof, it could as well be the plugin failing to return a data query response. Like it depends on how the plugin is implemented and what statuses decides to return. Or am I thinking wrongly about it?

**Review by marefr — COMMENTED** [review]
Some comments. Still many questions/concerns regarding this

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go (hunk 1 above)
Comment (marefr) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go]: Think my main concern with all this is that we would have more errorSources than just Grafana and downstream/plugin. From the perspective of defining SLO's for a team owning a plugin my thinking is that Grafana and downstream/plugin should be enough to understand if something is wrong with a certain plugin. But I guess the problem is that SLO's for those owning the plugin need to exclude errors happening downstream - so basically the same problem Grafana Cloud operators have today :exploding_head: 

Based on this I've come to the conclusion that each plugin should define their own custom metrics to be used for SLO purposes. Eager to hear your thoughts regarding this

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go [hunk 8]
```diff
@@ -58,29 +58,127 @@ const (
...
+}
+
+func InstrumentQueryDataRequest(ctx context.Context, cfg Cfg, pluginCtx *backend.PluginContext, requestSize float64, fn func() (*backend.QueryDataResponse, error)) (*backend.QueryDataResponse, error) {
+	start := time.Now()
+	timeBeforePluginRequest := log.TimeSinceStart(ctx, start)
+
+	resp, err := fn()
+
+	elapsed := time.Since(start)
+	logger.Info("Plugin Request Completed")
```
[the remark this answers is not in the document]
Comment (LudoVio) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go]: both fixed, thanks

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go [hunk 9]
```diff
@@ -58,29 +58,127 @@ const (
...
+func getStatus(err error) string {
+	if err == nil {
+		return statusOK
+	}
+
+	if errors.Is(err, context.Canceled) {
+		return statusCancelled
+	}
+
+	return statusError
```
[the remark this answers is not in the document]
Comment (LudoVio) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:169]: If any of those error happen, then the function `getStatus` would return `statusError`, and if there is a `statusError` then the function `getErrorSource` would return "pluginSource".

Is it not the expected behavior? Unless I made a mistake in the implementation somewhere

[the remark this answers is not in the document]
Comment (LudoVio) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:169]: I think I see what you mean, but the function that is being instrumented is the plugin one, not the client one. So the error that we got in the `instrumentation.go` are always the one from the plugin `p.QueryData(ctx, req)`.

`service.QueryData()` can return error from both plugin or Grafana, but that is not the function being instrumented

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go [hunk 10]
```diff
@@ -58,29 +58,127 @@ const (
...
+
+// Default implementation.
+// TODO: Provide a way to override this on a per-plugin basis. (need to modify plugin-sdk-go)
+func getErrorSource(status string, resp *backend.QueryDataResponse) errorSource {
+	if status == statusError {
+		return pluginSource
+	}
+
+	if status == statusCancelled {
+		return externalSource
```
[the remark this answers is not in the document]
Comment (LudoVio) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go]: Changed to `noneSource` then. That is an error that we probably always want to filter out anyway

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go (hunk 5 above)
Comment (LudoVio) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:121]: Yes, for the server point of view, 400 means that the client sent invalid data, but we (plugin) are the client here.
I set it to `pluginSource`, but we will probably fine tune that using logs and traces, because those 400 could come from the plugin, who made an error contructing the data, or it might come from the Grafana frontend, which is also owned by us. If the UI that allow this invalid data is from the plugin, then it is pluginSource, if it is from a global Grafana component somewhere, then it could be grafanaSource.

I think defaulting to pluginSource to begin with will allow us to get all our issues, plus other, so it is easier to dispatch in a later step

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go (hunk 7 above)
Comment (LudoVio) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:112]: You are right that the entire thing depends on how well the database implemented their http api, particularly the http status code.
I seems pretty good for Prometheus, but I know some other plugin would need a custom implementation of `getErrorSource` that is not based on http code.

For the 500, in theory it is supposed to be the database that crashed because of an issue from their side.
In practice there is a lot of services that return 500 instead of 400 if you don't send them valid data.

I don't think that there is a bulletproof solution, this one included. The idea is to first try with something that in theory, most of the time, should report the right thing, and then when we start receiving the data, and see manually if that is useful or not etc, then we iterate and improve the logic
